load/dependencies.go

```diff
@@ -7,24 +7,22 @@ import (
 	"context"
 )
 
-type TxGenerator[T comparable] interface {
-	// GenerateTx returns a valid transaction.
-	GenerateTx(ctx context.Context) (T, error)
+type Issuer[T comparable] interface {
+	// GenerateAndIssueTx generates and sends a tx to the network, and informs the
+	// tracker that it sent said transaction. It returns the sent transaction.
+	GenerateAndIssueTx(ctx context.Context) (tx T, err error)
 }
 
-type Issuer[T comparable] interface {
+type Listener[T comparable] interface {
 	// Listen for the final status of transactions and notify the tracker
-	// Listen stops if the context is done, an error occurs, or if the issuer
-	// has sent all their transactions.
+	// Listen stops if the context is done, an error occurs, or if it received
+	// all the transactions issued and the issuer no longer issues any.
+	// Listen MUST return a nil error if the context is canceled.
 	Listen(ctx context.Context) error
 
-	// Stop notifies the issuer that no further transactions will be issued.
-	// If a transaction is issued after Stop has been called, the issuer should error.
-	Stop()
-
-	// Issue sends a tx to the network, and informs the tracker that its sent
-	// said transaction.
-	IssueTx(ctx context.Context, tx T) error
+	// RegisterIssued informs the listener that a transaction was issued,
+	// and specifies if it was the last transaction to be issued.
+	RegisterIssued(tx T, last bool)
```

Comment:
Could we create a separate function to signal `last`? It seems that we're using a boolean to combine two distinct steps into a single function

Reply:
Added a `Listener` method `IssuingDone()` for this (similar to how it was with `Stop()`)

Reply:
Could we just keep the original name `Stop` ?

Reply:
IMO `Stop` is misleading since it does not actually stop the listener (at least in a blocking way), it tells the listener to start a timer to wait for the last outstanding transactions and then stop. Any other idea is welcome though